python_scripts/dev_tree.py

```diff
@@ -0,0 +1,783 @@
+)
+
+# %%
+data_reg_columns = ["Flipper Length (mm)"]
+target_reg_column = "Body Mass (g)"
+
+data_reg = data[data_reg_columns + [target_reg_column]]
+data_reg = data_reg.dropna()
+
+X_reg, y_reg = data_reg[data_reg_columns], data_reg[target_reg_column]
+X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
+    X_reg, y_reg, random_state=0,
+)
+
+# %%
+_, axs = plt.subplots(ncols=2, figsize=(10, 5))
+sns.scatterplot(
+    data=data_clf,
+    x="Culmen Length (mm)", y="Culmen Depth (mm)", hue="Species",
+    ax=axs[0],
+)
+axs[0].set_title("Classification dataset")
+sns.scatterplot(
+    data=data_reg, x="Flipper Length (mm)", y="Body Mass (g)",
+    ax=axs[1],
+)
+_ = axs[1].set_title("Regression dataset")
+
+# %% [markdown]
+# ### Effect of the `max_depth` parameter
+#
+# In decision tree, the most important parameter to get a trade-off between
+# under-fitting and over-fitting is the `max_depth` parameter. Let's build
+# a shallow tree (for both classification and regression) and a deeper tree.
+
+
+# %%
+max_depth = 2
+tree_clf = DecisionTreeClassifier(max_depth=max_depth)
+tree_reg = DecisionTreeRegressor(max_depth=max_depth)
+
+fig, axs = plt.subplots(ncols=2, figsize=(10, 5))
+plot_decision_function(X_train_clf, y_train_clf, tree_clf, ax=axs[0])
+plot_regression_model(X_train_reg, y_train_reg, tree_reg, ax=axs[1])
+_ = fig.suptitle(f"Shallow tree with a max-depth of {max_depth}")
+
+
+# %%
+max_depth = 30
+tree_clf.set_params(max_depth=max_depth)
+tree_reg.set_params(max_depth=max_depth)
+
+fig, axs = plt.subplots(ncols=2, figsize=(10, 5))
+plot_decision_function(X_train_clf, y_train_clf, tree_clf, ax=axs[0])
+plot_regression_model(X_train_reg, y_train_reg, tree_reg, ax=axs[1])
+_ = fig.suptitle(f"Deep tree with a max-depth of {max_depth}")
+
+# %% [markdown]
+# In both classification and regression setting, we can observe that increasing
+# the depth will make the tree model more expressive. However, a tree which is
+# too deep will overfit the training data, creating partitions which will only
+# be correct for "outliers". The `max_depth` is one of the parameter that one
+# would like to optimize via cross-validation and a grid-search.
+
+# %%
+from sklearn.model_selection import GridSearchCV
+
+param_grid = {"max_depth": np.arange(2, 10, 1)}
+tree_clf = GridSearchCV(DecisionTreeClassifier(), param_grid=param_grid)
+tree_reg = GridSearchCV(DecisionTreeRegressor(), param_grid=param_grid)
+
+# %%
+fig, axs = plt.subplots(ncols=2, figsize=(10, 5))
+plot_decision_function(X_train_clf, y_train_clf, tree_clf, ax=axs[0])
+axs[0].set_title(
+    f"Optimal depth found via CV: {tree_clf.best_params_['max_depth']}"
+)
+plot_regression_model(X_train_reg, y_train_reg, tree_reg, ax=axs[1])
+_ = axs[1].set_title(
+    f"Optimal depth found via CV: {tree_reg.best_params_['max_depth']}"
+)
+
+# %% [markdown]
+# The other parameters are used to fine tune the decision tree and have less
+# impact than `max_depth`.
```

Comment:
We could add a summary here, as in the previous notebook

Reply:
Yes, and probably something on the top of the notebook as well. I am just wondering if there is some missing aspects regarding the parameters to add thought